diff --git a/cpp/cmake/thirdparty/get_fmt.cmake b/cpp/cmake/thirdparty/get_fmt.cmake
index 083dd1d063..628588ccd9 100644
--- a/cpp/cmake/thirdparty/get_fmt.cmake
+++ b/cpp/cmake/thirdparty/get_fmt.cmake
@@ -16,7 +16,7 @@
 function(find_and_configure_fmt)
 
   include(${rapids-cmake-dir}/cpm/fmt.cmake)
-  rapids_cpm_fmt(INSTALL_EXPORT_SET cudf-exports BUILD_EXPORT_SET cudf-exports)
+  rapids_cpm_fmt(INSTALL_EXPORT_SET cudf-exports BUILD_EXPORT_SET cudf-exports EXCLUDE_FROM_ALL)
 endfunction()
 