--- a/folly/portability/Builtins.h	2026-04-16 10:25:24.897617900 -0700
+++ b/folly/portability/Builtins.h	2026-04-16 10:25:25.048394000 -0700
@@ -135,6 +135,15 @@
   return int(__popcnt(x));
 }
 #endif // !defined(_MSC_VER) || !defined(FOLLY_DETAIL_MSC_BUILTIN_SUPPORT)
+
+#if defined(_M_ARM) || defined(_M_ARM64)
+FOLLY_ALWAYS_INLINE int __builtin_popcount(unsigned int x) {
+  return int(_CountOneBits(static_cast<unsigned long>(x)));
+}
+FOLLY_ALWAYS_INLINE int __builtin_popcountl(unsigned long x) {
+  return int(_CountOneBits(x));
+}
+#endif
 #endif
 
 #if !defined(_MSC_VER) || !defined(FOLLY_DETAIL_MSC_BUILTIN_SUPPORT)
@@ -147,6 +156,11 @@
 FOLLY_ALWAYS_INLINE int __builtin_popcountll(unsigned long long x) {
   return int(__popcnt64(x));
 }
+#elif defined(_M_ARM64)
+FOLLY_ALWAYS_INLINE int __builtin_popcountll(unsigned long long x) {
+  return int(_CountOneBits(static_cast<unsigned long>(x >> 32))) +
+      int(_CountOneBits(static_cast<unsigned long>(x)));
+}
 #endif
 #endif // !defined(_MSC_VER) || !defined(FOLLY_DETAIL_MSC_BUILTIN_SUPPORT)
 
