diff --git a/cpp/src/arrow/util/bit_util.h b/cpp/src/arrow/util/bit_util.h
index d30baaac65..585d1e9e7e 100644
--- a/cpp/src/arrow/util/bit_util.h
+++ b/cpp/src/arrow/util/bit_util.h
@@ -21,12 +21,23 @@
 #if defined(_M_AMD64) || defined(_M_X64)
 #include <intrin.h>  // IWYU pragma: keep
 #include <nmmintrin.h>
-#endif
-
 #pragma intrinsic(_BitScanReverse)
 #pragma intrinsic(_BitScanForward)
 #define ARROW_POPCOUNT64 __popcnt64
 #define ARROW_POPCOUNT32 __popcnt
+#elif defined(_M_ARM64)
+inline unsigned __int64 ARROW_POPCOUNT64(unsigned __int64 x) {
+    unsigned __int64 c = 0;
+    for(; x; c++) x &= x - 1;
+    return c;
+}
+inline unsigned int ARROW_POPCOUNT32(unsigned int x) {
+    unsigned int c = 0;
+    for(; x; c++) x &= x - 1;
+    return c;
+}
+#endif
+
 #else
 #define ARROW_POPCOUNT64 __builtin_popcountll
 #define ARROW_POPCOUNT32 __builtin_popcount
@@ -174,12 +185,12 @@ static inline uint64_t TrailingBits(uint64_t v, int num_bits) {
 }
 
 /// \brief Count the number of leading zeros in an unsigned integer.
-static constexpr inline int CountLeadingZeros(uint32_t value) {
+static inline int CountLeadingZeros(uint32_t value) {
 #if defined(__clang__) || defined(__GNUC__)
   if (value == 0) return 32;
   return static_cast<int>(__builtin_clz(value));
 #elif defined(_MSC_VER)
-  unsigned long index;                                               // NOLINT
+  unsigned long index = 0;                                           // NOLINT
   if (_BitScanReverse(&index, static_cast<unsigned long>(value))) {  // NOLINT
     return 31 - static_cast<int>(index);
   } else {
@@ -200,7 +211,7 @@ static inline int CountLeadingZeros(uint64_t value) {
   if (value == 0) return 64;
   return static_cast<int>(__builtin_clzll(value));
 #elif defined(_MSC_VER)
-  unsigned long index;                     // NOLINT
+  unsigned long index = 0;                 // NOLINT
   if (_BitScanReverse64(&index, value)) {  // NOLINT
     return 63 - static_cast<int>(index);
   } else {
