diff --git a/cpp/src/arrow/device.h b/cpp/src/arrow/device.h
index efb0a5ab40..876397ed14 100644
--- a/cpp/src/arrow/device.h
+++ b/cpp/src/arrow/device.h
@@ -139,7 +139,7 @@ class ARROW_EXPORT Device : public std::enable_shared_from_this<Device>,
   /// This should create the appropriate stream type for the device,
   /// derived from Device::Stream to allow for stream ordered events
   /// and memory allocations.
-  virtual Result<std::shared_ptr<Stream>> MakeStream(unsigned int flags) {
+  virtual Result<std::shared_ptr<Stream>> MakeStream(unsigned int) {
     return NULLPTR;
   }
 
@@ -149,8 +149,8 @@ class ARROW_EXPORT Device : public std::enable_shared_from_this<Device>,
   /// @param release_fn a function to call during destruction, `nullptr` or
   ///        a no-op function can be passed to indicate ownership is maintained
   ///        externally
-  virtual Result<std::shared_ptr<Stream>> WrapStream(void* device_stream,
-                                                     Stream::release_fn_t release_fn) {
+  virtual Result<std::shared_ptr<Stream>> WrapStream(void*,
+                                                     Stream::release_fn_t) {
     return NULLPTR;
   }
 
