--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -36,27 +36,6 @@
 set(CMAKE_VERBOSE_MAKEFILE OFF)
 set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 set(CMAKE_MACOSX_RPATH 1)
-
-if(NOT DEFINED CMAKE_C_COMPILER_LAUNCHER)
-    find_program(COMPILER_LAUNCHER NAMES ccache sccache)
-    if(COMPILER_LAUNCHER)
-        message(STATUS "Using ${COMPILER_LAUNCHER} as C compiler launcher")
-        set(CMAKE_C_COMPILER_LAUNCHER
-                "${COMPILER_LAUNCHER}"
-                CACHE STRING "" FORCE)
-    endif()
-endif()
-
-if(NOT DEFINED CMAKE_CXX_COMPILER_LAUNCHER)
-    find_program(COMPILER_LAUNCHER NAMES ccache sccache)
-    if(COMPILER_LAUNCHER)
-        message(STATUS "Using ${COMPILER_LAUNCHER} as C++ compiler launcher")
-        set(CMAKE_CXX_COMPILER_LAUNCHER
-                "${COMPILER_LAUNCHER}"
-                CACHE STRING "" FORCE)
-    endif()
-endif()
-
 # Determine install paths
 # default to gnu standard installation directories (lib, bin, include)
 # https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html
