--- a/third_party/zstd/CMakeLists.txt
+++ b/third_party/zstd/CMakeLists.txt
@@ -40,7 +40,7 @@ add_library(duckdb_zstd STATIC ${ZSTD_FILES})

 target_include_directories(
 		duckdb_zstd
-		PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
+		PRIVATE $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
 set_target_properties(duckdb_zstd PROPERTIES EXPORT_NAME duckdb_duckdb_zstd)

 install(TARGETS duckdb_zstd
