--- a/src/s2/s2cell_id.cc
+++ b/src/s2/s2cell_id.cc
@@ -195,7 +195,7 @@ int S2CellId::GetCommonAncestorLevel(S2CellId other) const {
   // Compute the position of the most significant bit, and then map the bit
   // position as follows:
   // {0} -> 30, {1,2} -> 29, {3,4} -> 28, ... , {59,60} -> 0, {61,62,63} -> -1.
-  return max(61 - absl::bit_width(bits), -1) >> 1;
+  return std::max<int>(61 - absl::bit_width(bits), -1) >> 1;
 }

 // Print the num_digits low order hex digits.
